docs/DesignDocs/Enabledebugmalloc.md

```diff
@@ -0,0 +1,154 @@
+identify leaks within the enclave. Commonly used memory detection tools like Valgrind do not
+work within the enclaves since they cannot access enclave memory.
+
+# Design
+
+## User Experience
+- Debug malloc will be shipped to the users as a separate library oelibdebugmalloc.a
+- To use debug malloc, user will have to link oelibdebugmalloc.a into their enclaves.
+  This can be done by
+  ```
+      cmake:
+      target_link_libraries(user-enclave oelibc oedebugmalloc.a oeenclave)
+
+      make:
+      LDFLAGS=$(shell pkg-config oeenclave-$(COMPILER) --libs)
+      LDFLAGS_CUSTOM=$(subst -loelibc, -loedebugmalloc -loelibc, $(LDFLAGS))
+      $(CC) -o enclave enc.o $(LDFLAGS_CUSTOM)
+
+  ```
+
+## Fine-grained runtime control
+Currently, debug malloc prints leaks only during enclave termination.
+However, users would like to isolate and track specific memory allocations.
+To achieve this, the following APIs are provided in a new published header `debugmalloc.h`.
+
+- `void oe_debug_malloc_start_tracking(void)`
+  Start tracking memory allocations.
+
+- `void oe_debug_malloc_stop_tracking(void)`
+  Stop tracking memory allocations.
+
+- `void oe_debug_malloc_print_objects(void)`
+  Print tracked objects that alive, ie not yet freed.
+
+- `void oe_debug_malloc_init_callback(void)`
+  This function will be called by debug malloc during startup.
+  Users can implement this function to start a tracking session at the earliest possible
+  time.
+
+Please note that the timing of calling of these functions is important. For
+example, the starting and stoping tracking must cover the coupled malloc/free
+operations, which means start tracking before malloc and also stop tracking
+after free, or debug malloc will definitely trigger false alarms. And by the same
+logic, oe_debug_malloc_print_objects() can only be called after tracking is
+stopped.
+
+# Implementation
+
+## Building OE SDK from source retains current behavior.
+
+OE SDK build system will retain the USE_DEBUG_MALLOC cmake parameter.
+By default, in debug build, it is set to `on` and enclaves will link against
+oe_debug_malloc and enable leak detection.
+
+The default implementation of oe_debug_malloc_init_callback will be thus:
+```c
+void oe_debug_malloc_init_callback(void)
+{
+#ifdef OE_USE_DEBUG_MALLOC // defined when USE_DEBUG_MALOC=on
+   oe_debug_malloc_start_tracking();
+#endif
+}
+```
+Of course, this default setting can be turned off by changing USE_DEBUG_MALLOC
+to `off`.
+
+## User-defined callback function to subsitute the predefined function
+While user defines the callback function following the definition and matches
+the name exactly, the predefined function will be substituted by user's, thus
+user can decide the point to start tracking by calling the callback function.
+
+The expected implementation of the callback function is as following:
+  ```c
+  void oe_debug_malloc_init_callback(void)
```

Comment:
As mentioned in the SIG arch meeting, I would prefer if we didn't offer an init callback method:

- It provides a lot of freedom for a dev to implement code this is unsafe or incorrect, given that the callback is invoked at a point that is heavily constrained during enclave init. 
- The primary use case for this seems to be to allow a developer to start tracking at the point of allocator init, and there are simpler mechanisms for that.

Reply:
Thanks! callback is removed.